dask/dataframe/io/io.py

```diff
@@ -740,5 +747,145 @@ def sorted_division_locations(seq, npartitions=None, chunksize=None):
     return values, positions
 
 
+@insert_meta_param_description
+def from_map(
+    func,
+    inputs,
+    meta=None,
+    divisions=None,
+    label=None,
+    token=None,
+    enforce_metadata=True,
+    **kwargs,
+):
+    """Create a DataFrame collection from a custom function map
+
+    Parameters
+    ----------
+    func : function
+        Function used to create each partition.
+    inputs : Iterable
+        Iterable object of arguments to pass to ``func``. The number of
+        elements in ``inputs`` will correspond to the number of partitions
+        in the output collection (only one element will be passed to
+        ``func`` for each partition).
+    $META
+    divisions : tuple, str, optional
+        Partition boundaries along the index.
+        For tuple, see https://docs.dask.org/en/latest/dataframe-design.html#partitions
+        For string 'sorted' will compute the delayed values to find index
+        values.  Assumes that the indexes are mutually sorted.
+        If None, then won't use index information
+    label : str, optional
+        String to use as the function-name label in the output
+        collection-key names.
+    token : str, optional
+        String to use as the "token" in the output collection-key names.
+    enforce_metadata : bool, default True
+        Whether to enforce at runtime that the structure of the DataFrame
+        produced by ``func`` actually matches the structure of ``meta``.
+        This will rename and reorder columns for each partition,
+        and will raise an error if this doesn't work or types don't match.
+    **kwargs:
+        Dictionary of key-word arguments to be passed to ``func`` for every
+        output partition.
```

Comment:
One thing that SciPy does in a few places is provide an `args` keyword argument so you can pass positional arguments in addition to `kwargs, like:

```python
def func(data, arg1, arg2):
     pass

dd.from_map(
    func,
    iter,
    args=(arg1_value, arg2_value),
)
```

Do you think it makes sense to add something like that here? `kwargs` is more general, but positional args are useful sometimes too, especially if you're integrating with an "third-party" API that you don't control or don't want to write a wrapper for.

Reply:
>Do you think it makes sense to add something like that here? kwargs is more general, but positional args are useful sometimes too, especially if you're integrating with an "third-party" API that you don't control or don't want to write a wrapper for.

This request make a lot of sense, but I'm going to push back unless you (or others) feel strongly. I fear that it would be a bit more confusing than helpful to add `args`. I'm hoping to keep the scope of this API as simple as possible, and if we added an `args=` argument, then we would need to place an arbitrary requirement on how the `args` should be ordered with respect to `inputs`.  We could certainly work this all out (and document it), but in the end the user can always pass in a positional argument as a key-word argument in modern Python versions. For example, my understanding is that the following would work fine for your example:

```python
dd.from_map(
    func,
    iter,
    arg1=arg1_value,
    arg2=arg2_value,
)
```

Reply:
I see that @bryanwweber bumped on this as well, though I came to the conclusion that we should just unpack `*inputs` in the function directly, mostly for consistency with dask's other map-ish functions

Reply:
Thanks @rjzamora, I think that makes sense and I don't see this as a blocker here, more of a question. I think it would need to be added to _all_ the similar APIs that already exist, so it'd be quite a lot of work.

> Example..

That would work as long as the args aren't positional only, which is the case for certain C-API functions depending on how they're written. I suspect that's why SciPy supports it, the cases I know it from are where SciPy is wrapping a solver in C/Fortran.

As far as the ordering with respect to the inputs, I believe the way it's handled is all the `args` go right after the `data`, followed by all the `kwargs`.

Reply:
>I see that @bryanwweber bumped on this as well, though I came to the conclusion that we should just unpack *inputs in the function directly, mostly for consistency with dask's other map-ish functions

Not sure I understand exactly what you mean by this @ian-r-rose - Are you suggesting that the user could pass in an arbitrary number of iterables, and inter-mix it with non-iterable inputs that would be broadcasted to all partitions? Or are you saying that we should just allow an arbitrary number of iterable arguments?

It seems like it would be tricky to inter-mix the positional inputs, since we cannot simply check which arguments are `Iterable` in the same way `map_partitions` can check if the inputs are collections. It seems reasonable for the user to pass in an Iterable that they actually want to broadcast.

>That would work as long as the args aren't positional only, which is the case for certain C-API functions depending on how they're written. I suspect that's why SciPy supports it, the cases I know it from are where SciPy is wrapping a solver in C/Fortran.

I see - That's a failr point.

>As far as the ordering with respect to the inputs, I believe the way it's handled is all the args go right after the data, followed by all the kwargs.

This ordering definitely makes the most intuitive sense to me. However, we will probably need to build a small wrapper class to handle it, because something like `functools.partial` only allows you to freeze the leading postional arguments (I *think*) :/

Reply:
> Not sure I understand exactly what you mean by this @ian-r-rose - Are you suggesting that the user could pass in an arbitrary number of iterables, and inter-mix it with non-iterable inputs that would be broadcasted to all partitions? Or are you saying that we should just allow an arbitrary number of iterable arguments?

What I wrote above didn't make much sense, apologies. What I'm thinking about now is matching something like what `Client.map` does:

```python
def from_map(
    func: Callable,
    *inputs: Iterable | BlockwiseDep
     ...
    ***kwargs
)
```
with the constraint that each of `inputs` must be of the same length (including possible `BlockwiseDep` inputs, which would be broadcast). I think any non-iterable-non-`BlockwiseDep` input would be invalid.

Reply:
I think I agree with the idea to handle multiple Iterables, but I'd actually prefer not to support `BlockwiseDep` here (which may or may not support `len`). My (not-so-strong) opinion is that someone working with a `BlockwiseDep` object is probably enough of a super-user to construct a Blockwise layer manually :)

Reply:
That's fair. I *almost* wish that `map_partitions` didn't handle `BlockwiseDep` for the same reason.

Reply:
> > That would work as long as the args aren't positional only, which is the case for certain C-API functions depending on how they're written. I suspect that's why SciPy supports it, the cases I know it from are where SciPy is wrapping a solver in C/Fortran.
> 
> I see - That's a failr point.

Like I said, I don't see this as a blocker for this PR 😄 I'd definitely defer adding this until someone asks for it.
 
> This ordering definitely makes the most intuitive sense to me. However, we will probably need to build a small wrapper class to handle it, because something like `functools.partial` only allows you to freeze the leading postional arguments (I _think_) :/

Yeah, another reason to defer adding this 😄

Reply:
I did end up adding the `args=` option since I already needed a wrapper class to handle column projection. So, I'll mark this discussion as "resolved," but feel free to comment further.